ksql-engine/src/main/java/io/confluent/ksql/function/FunctionNameValidator.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.ksql.function;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * Check that a function name is valid. It is valid if it is not a Java reserved word
+ * and not a ksql reserved word and is a valid java identifier.
+ */
+class FunctionNameValidator implements Predicate<String> {
+  private static final Set<String> JAVA_RESERVED_WORDS
+      = ImmutableSet.<String>builder()
+      .add("abstract").add("assert").add("boolean").add("break").add("byte").add("case")
+      .add("catch").add("char").add("class").add("const").add("continue").add("default")
+      .add("double").add("else").add("enum").add("extends").add("do").add("final").add("finally")
+      .add("float").add("for").add("goto").add("if").add("int").add("implements").add("import")
+      .add("instanceof").add("interface").add("long").add("native").add("new").add("package")
+      .add("private").add("public").add("protected").add("return").add("this").add("throw")
+      .add("throws").add("transient").add("try").add("short").add("static").add("strictfp")
+      .add("super").add("switch").add("synchronized").add("void").add("volatile").add("while")
+      .build();
+
+  // Anything else that should be added to this list?
+  private static final Set<String> KSQL_RESERVED_WORDS
+      = ImmutableSet.<String>builder()
+      .add("analyze").add("as").add("catalog").add("create").add("beginning").add("by")
+      .add("delete").add("describe").add("drop").add("exists").add("explain").add("export")
+      .add("extended").add("insert").add("into").add("join").add("from").add("list").add("load")
+      .add("limit").add("not").add("partition").add("print").add("properties").add("queries")
+      .add("registered").add("run").add("script").add("select").add("set").add("show")
+      .add("stream").add("streams").add("table").add("tables").add("terminate")
+      .add("topics").add("unset").add("when").add("where").add("window").add("with")
+      .build();
```

Comment:
Seems that some of the reserved words such as `GROUP`, `LEFT' and `CAST` are missing. You can find all of them in the grammar or even a better source would be the generated token file.

Reply:
Yeah - that is why i left the comment. Where does the token file get generated to?

Reply:
found it

Reply:
feels very brittle ? normally the generated SqlBaseParser.java class will contain a map of all the grammar symbols - is that something we could use dynamically here ?

Reply:
ah. I just saw this thread. Yes, the lexical tokens are available in the code through the `SqlBaseParser.vocabulary` field. We should just use that.